streaming/replication.c

```diff
@@ -0,0 +1,1931 @@
+        close(rep_state->socket);
+        return 0;
+    }
+    
+    // call the plugins.d processor to receive the metrics
+    info("%s %s [receive from [%s]:%s]: filling replication gaps...", REPLICATION_MSG, host->hostname, rep_state->client_ip, rep_state->client_port);
+    log_replication_connection(rep_state->client_ip, rep_state->client_port, rep_state->key, host->machine_guid, host->hostname, "CONNECTED");
+
+    cd.version = rep_state->stream_version;
+
+    // Wait for the sender thread to send REP ON
+    size_t count = replication_parser(rep_state, &cd, fp);
+
+    // Replication completed or exitted
+    replication_thread_close_socket(rep_state);
+    
+    log_replication_connection(rep_state->client_ip, rep_state->client_port, rep_state->key, host->machine_guid, host->hostname, "DISCONNECTED");
+    debug(D_REPLICATION, "%s: %s [receive from [%s]:%s]: disconnected (completed %zu updates).", REPLICATION_MSG, host->hostname, rep_state->client_ip, rep_state->client_port, count);
+
+    info("%s: Replication Parser Finished (completed %zu updates)!", REPLICATION_MSG, count);
+    fclose(fp);
+    netdata_thread_cleanup_pop(1);
+    return NULL;   
+}
+
+int finish_gap_replication(RRDHOST *host, REPLICATION_STATE *rep_state) {
+    int num_of_queued_gaps = host->gaps_timeline->gaps->count;
+    if(!num_of_queued_gaps) {
+        info("%s: No more GAPs to replicate for host %s. Switch off the Replication Rx thread", REPLICATION_MSG, host->hostname);
+        // Send REP OFF to terminate replication at the Tx side.
+        send_message(rep_state, "REP 1\n");
+        rep_state->shutdown = 1;
+        // Start replication sender thread (Tx) for child image hosts. After the GAP replication. Bottom-Up approach.
+        if (host->replication->tx_replication->enabled && host->replication->tx_replication->spawned) {
+            info("%s: No more GAPs to replicate for host %s. Switch ON the Replication Tx thread to send the metrics to higher level hops.",
+                REPLICATION_MSG,
+                host->hostname);
+            host->replication->tx_replication->pause = 0;
+            host->replication->tx_replication->resume = 1;
+        }
+        return 1;
+    }
+    return 0;
+}
+
+void send_gap_for_replication(RRDHOST *host, REPLICATION_STATE *rep_state)
+{
+    GAP *the_gap = (GAP *)host->gaps_timeline->gaps->front->item;
+    //  Assign the timestamp of first metric comes from streaming to avoid
+    // missing metrics between the disconnection and start of the streaming
+    // the_gap->t_window.t_end = rep_state->host->receiver->first_msg_t - 1;
+    the_gap->t_window.t_end = rep_state->host->receiver->first_msg_t;
+    char *rep_msg_cmd;
+    size_t len;
+    replication_gap_to_str(the_gap, &rep_msg_cmd, &len);
+    send_message(rep_state, rep_msg_cmd);
+    the_gap->status = "ontransmit";
```

Comment:
I remember to do reverse engineer in codes generated by ancient compilers where strings we added to stack instead to have a static variable defined, if somehow we compile this code with these compilers, the status could be completely  unexpected. I suggest to use this `ontransmit` as a definition or a static variable to avoid this.